From 4e28bdb801c5063b2e29f354bb9b3ea6abb04d4b Mon Sep 17 00:00:00 2001
From: Jianxu Zheng <jian.xu.zheng@intel.com>
Date: Sat, 14 Feb 2015 11:56:16 +0800
Subject: [PATCH 272/273] ov8858: fix klockwork issue

in ov8858 driver v4l2_subdev *sd is not checked if it's null.
although it's a convention in ov8858 driver that all of this pointer
is not checked.

klockwork logs:
Suspicious dereference of pointer 'sd' by passing argument 1
to function 'v4l2_get_subdevdata' at line 741 before NULL
check at line 744
File:
/workspace1/AndroidKW100_imin_legacy_mofd_v1_64/SourceCode/
linux/modules/camera/drivers/media/i2c/ov8858.c:741
| __gpio_ctrl()

https://jira01.devtools.intel.com/browse/IMINAN-20360
Change-Id: Ibcec1f80adf9b104ab98de560408065af30d0a34
Signed-off-by: Jianxu Zheng <jian.xu.zheng@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/41520
Reviewed-by: Soren Friis <soren.friis@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Integrator: Wei Tao <wei.tao@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 drivers/external_drivers/camera/drivers/media/i2c/ov8858.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c b/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
index 4e98be7..2b69192 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
@@ -738,10 +738,16 @@ static int __power_ctrl(struct v4l2_subdev *sd, bool flag)
 
 static int __gpio_ctrl(struct v4l2_subdev *sd, bool flag)
 {
-	struct i2c_client *client = v4l2_get_subdevdata(sd);
-	struct ov8858_device *dev = to_ov8858_sensor(sd);
+	struct i2c_client *client;
+	struct ov8858_device *dev;
 
-	if (!dev || !dev->platform_data)
+	if (!sd)
+		return -EINVAL;
+
+	client = v4l2_get_subdevdata(sd);
+	dev = to_ov8858_sensor(sd);
+
+	if (!client || !dev || !dev->platform_data)
 		return -ENODEV;
 
 	/* Non-gmin platforms use the legacy callback */
-- 
1.9.1

